### PR TITLE
frida-inject send stdin to script

### PR DESCRIPTION
**Pull request by WorksButNotTested**

### CODE REVIEW

File: inject/inject.vala [hunk 1]
```diff
@@ -439,6 +474,15 @@ namespace Frida.Inject {
...
 		}
 
+		public async void call_send (string message) {
+			var msg = new Json.Node.alloc ().init_string (message);
+
+			try {
+				yield rpc_client.call ("recv", new Json.Node[] { msg }, io_cancellable);
+			} catch (GLib.Error e) {
+			}
+		}
```
Comment (oleavr) [inline inject/inject.vala:484]: I'd put this after `call_init()` as it's called after it. Or at least we should probably guarantee that it is.

File: inject/inject.vala [hunk 2]
```diff
@@ -225,6 +225,41 @@ namespace Frida.Inject {
 
 			exit_code = 0;
 
+			var fd = stdin.fileno();
```
Comment (oleavr) [inline inject/inject.vala:228]:
```suggestion
			var fd = stdin.fileno ();
```

File: inject/inject.vala [hunk 3]
```diff
@@ -225,6 +225,41 @@ namespace Frida.Inject {
...
+#if WINDOWS
+			var inchan = new IOChannel.win32_new_fd (fd);
+#else
+			var inchan = new IOChannel.unix_new (fd);
+#endif
+			inchan.add_watch (IOCondition.IN, (source, condition) => {
+				string str_return = null;
```
Comment (oleavr) [inline inject/inject.vala:235]: I'd move this to right before `try {` so it makes it clear what's about to happen inside the `try`.

File: inject/inject.vala [hunk 4]
```diff
@@ -225,6 +225,41 @@ namespace Frida.Inject {
...
+				size_t length = -1;
+				size_t terminator_pos = -1;
```
Comment (oleavr) [inline inject/inject.vala:237]: These could be moved inside the `try` as they're unused outside it. (Actually I see now that they're nullable at the C API level, so would be better to pass `null`, assuming the vapi is correct.)

File: inject/inject.vala [hunk 5]
```diff
@@ -225,6 +225,41 @@ namespace Frida.Inject {
...
+				}
+
+				try {
+					IOStatus status = inchan.read_line (out str_return, out length, out terminator_pos);
+					if (status == IOStatus.EOF) {
+						loop.quit ();
+						return false;
+					}
+
+					Idle.add(() => {
```
Comment (oleavr) [inline inject/inject.vala:250]:
```suggestion
					Idle.add (() => {
```

File: inject/inject.vala [hunk 6]
```diff
@@ -225,6 +225,41 @@ namespace Frida.Inject {
...
+						script_runner.call_send.begin (str_return);
+						return false;
+					});
```
Comment (oleavr) [inline inject/inject.vala:253]: This should be moved outside the `try` as it doesn't throw.

**Review by oleavr — CHANGES_REQUESTED** [review]
Nice! 🚀

File: inject/inject.vala (hunk 3 above)
Comment (WorksButNotTested) [reply inject/inject.vala:235]: I don't think I can move this? `inchan` is used to call `add_watch` so it needs to exist outside the callback. The try needs to handle potential errors occurring inside the callback.

Comment (oleavr) [reply inject/inject.vala:249]: I meant the `str_return` variable, sorry for being unclear :)

Comment (WorksButNotTested) [reply inject/inject.vala:249]: Ok thanks.

**Comment by WorksButNotTested** [comment]
Sample code
```javascript
function make_blue (message) {
    var escape = String.fromCharCode(parseInt('033', 8));
    var prefix = "[0;34m";
    var suffix = "[0m\n";
    var blue = `${escape}${prefix}${message.trim()}${escape}${suffix}`
    return blue;
}

rpc.exports = {
    onFridaStdIn: function (message) {
        var stdOutData = make_blue (`> ${message.trim()}`)
        send(['frida:stdout', stdOutData]);
    }
}

send(['frida:stdout', make_blue("STARTED\n")]);

var info = {
    type: "log",
    level: "info",
    payload: "> error message"
}
_send (JSON.stringify(info), null);

var warning = {
    type: "log",
    level: "warning",
    payload: "> warning message"
}
_send (JSON.stringify(warning), null);

var error = {
    type: "log",
    level: "error",
    payload: "> error message"
}
_send (JSON.stringify(error), null);
```

File: inject/inject.vala [hunk 7]
```diff
@@ -508,6 +573,56 @@ namespace Frida.Inject {
...
+		 * send(['frida:stderr', 'MESSAGE']);
+		 *
+		 * The resulting message will look as shown below. Note that we don't
+		 * use the parents object's `type` field since this is reserved for use
+		 * by the runtime itself.
+		 *
+		 * {"type":"send","payload":["frida:stdout","MESSAGE"]}
+		 */
+		private bool try_handle_stdout_message (Json.Object message) {
+			var payload = message.get_array_member ("payload");
```
Comment (oleavr) [inline inject/inject.vala:592]: This API requires it to actually be an array if it exists. Might be easier to use a `JsonReader`.

File: inject/inject.vala [hunk 8]
```diff
@@ -508,6 +573,56 @@ namespace Frida.Inject {
...
+			if (payload == null)
+				return false;
+
+			if (payload.get_length() != 2)
+				return false;
+
+			var type = payload.get_string_element (0);
```
Comment (oleavr) [inline inject/inject.vala:599]: Like above, this requires the element to be a string.

File: inject/inject.vala [hunk 9]
```diff
@@ -508,6 +573,56 @@ namespace Frida.Inject {
...
+		 * by the runtime itself.
+		 *
+		 * {"type":"send","payload":["frida:stdout","MESSAGE"]}
+		 */
+		private bool try_handle_stdout_message (Json.Object message) {
+			var payload = message.get_array_member ("payload");
+			if (payload == null)
+				return false;
+
+			if (payload.get_length() != 2)
```
Comment (oleavr) [inline inject/inject.vala:596]:
```suggestion
			if (payload.get_length () != 2)
```

File: inject/inject.vala [hunk 10]
```diff
@@ -508,6 +573,56 @@ namespace Frida.Inject {
...
+				return false;
+
+			var type = payload.get_string_element (0);
+			switch (type)
+			{
```
Comment (oleavr) [inline inject/inject.vala:601]: Opening brace on wrong line.

File: inject/inject.vala [hunk 11]
```diff
@@ -508,6 +573,56 @@ namespace Frida.Inject {
...
+				default:
+					return false;
+			}
+
+			var msg = payload.get_string_element (1);
+			if (msg == null)
+				return false;
+
+			switch (type)
+			{
```
Comment (oleavr) [inline inject/inject.vala:614]: Opening brace on wrong line.

File: inject/inject.vala [hunk 12]
```diff
@@ -225,6 +225,53 @@ namespace Frida.Inject {
...
+				}
+
+				string str_return = null;
+				try {
+					IOStatus status = inchan.read_line (out str_return, null, null);
+					if (status == IOStatus.EOF) {
+						loop.quit ();
+						return false;
+					}
+
```
Comment (oleavr) [inline inject/inject.vala:260]:
```suggestion
```

File: inject/inject.vala [hunk 13]
```diff
@@ -225,6 +225,53 @@ namespace Frida.Inject {
...
+			 * }
+			 */
+			var fd = stdin.fileno ();
+#if WINDOWS
+			var inchan = new IOChannel.win32_new_fd (fd);
+#else
+			var inchan = new IOChannel.unix_new (fd);
+#endif
+			inchan.add_watch (IOCondition.IN, (source, condition) => {
+				if (condition == IOCondition.HUP) {
```
Comment (oleavr) [inline inject/inject.vala:249]: Braces can be omitted.

File: inject/inject.vala [hunk 14]
```diff
@@ -225,6 +225,53 @@ namespace Frida.Inject {
...
+					return false;
+				}
+
+				string str_return = null;
+				try {
+					IOStatus status = inchan.read_line (out str_return, null, null);
+					if (status == IOStatus.EOF) {
```
Comment (oleavr) [inline inject/inject.vala:256]: Let's move this outside the `try`.

**Review by oleavr — CHANGES_REQUESTED** [review]

**Comment by WorksButNotTested** [comment]
function make_blue (message) {
    var escape = String.fromCharCode(parseInt('033', 8));
    var prefix = "[0;34m";
    var suffix = "[0m\n";
    var blue = `${escape}${prefix}${message.trim()}${escape}${suffix}`
    return blue;
}

rpc.exports = {
    onFridaStdIn: function (message) {
        var stdOutData = make_blue (`> ${message.trim()}`)
        send(['frida:stdout', stdOutData]);
    }
}

send(['frida:stdout', make_blue("STARTED\n")]);

var info = {
    type: "log",
    level: "info",
    payload: "> error message"
}
_send (JSON.stringify(info), null);

var warning = {
    type: "log",
    level: "warning",
    payload: "> warning message"
}
_send (JSON.stringify(warning), null);

var error = {
    type: "log",
    level: "error",
    payload: "> error message"
}
_send (JSON.stringify(error), null);

send ("BLAH");
send (["test1", "test2", "test3"]);
send ([123, "test"]);
send (["frida:stdout", 123]);
send ([{ key: 123, value: "test"}, "abc"]);
send (["frida:stdout", { key: 123, value: "test"}]);
send ({ key: 123, value: "test"});
send (null);

**Comment by WorksButNotTested** [comment]
The only input I can think which causes a crash now is when you call send without any parameters:
```javascript
send();
```

But then we don't even get as far as the `on_message` function.

**Comment by oleavr** [comment]
Thanks! 🙌

**Pull request merged**